File: ios/Sources/CblitePlugin.swift
import Foundation
import Tauri
import UIKit
import WebKit
import CouchbaseLiteSwift

// ---------------------------------------------------------------------------
// CblitePlugin — iOS implementation of tauri-plugin-cblite, backed by the
// CouchbaseLiteSwift SDK. Mirrors the Android Kotlin plugin command-for-command:
//   openDatabase, closeDatabase, getDocument, saveDocument, executeQuery (N1QL),
//   createFtsIndex, startReplication, stopReplication, saveBlob, getBlobData.
// Response shapes match Kotlin: queries resolve { rows: [...] }, blobs resolve
// { value: "..." }, documents resolve their JSON body directly.
// ---------------------------------------------------------------------------

// ── Arbitrary-JSON Decodable (for saveDocument bodies) ──────────────────────

enum JSONValue: Decodable {
    case null
    case bool(Bool)
    case number(Double)
    case string(String)
    case array([JSONValue])
    case object([String: JSONValue])

    init(from decoder: Decoder) throws {
        let c = try decoder.singleValueContainer()
        if c.decodeNil() {
            self = .null
        } else if let b = try? c.decode(Bool.self) {
            self = .bool(b)
        } else if let n = try? c.decode(Double.self) {
            self = .number(n)
        } else if let s = try? c.decode(String.self) {
            self = .string(s)
        } else if let a = try? c.decode([JSONValue].self) {
            self = .array(a)
        } else if let o = try? c.decode([String: JSONValue].self) {
            self = .object(o)
        } else {
            throw DecodingError.dataCorruptedError(in: c, debugDescription: "unsupported JSON value")
        }
    }

    /// Convert to a Foundation object suitable for JSONSerialization.
    var anyValue: Any {
        switch self {
        case .null: return NSNull()
        case .bool(let b): return b
        case .number(let n):
            // Render integral numbers as Int so JSON round-trips without ".0".
            if n.truncatingRemainder(dividingBy: 1) == 0, n >= -9_007_199_254_740_991, n <= 9_007_199_254_740_991 {
                return Int64(n)
            }
            return n
        case .string(let s): return s
        case .array(let a): return a.map { $0.anyValue }
        case .object(let o): return o.mapValues { $0.anyValue }
        }
    }
}

// ── Command argument types (mirror commands_mobile.rs payloads) ─────────────

class OpenDatabaseArgs: Decodable {
    let path: String
    let name: String
    var encryptionPassword: String? = nil
    var collections: [String]? = nil
}

class DocArgs: Decodable {
    let collection: String
    let docId: String
}

class SaveDocumentArgs: Decodable {
    let collection: String
    let docId: String
    let body: JSONValue
    var encryptedFields: [String]? = nil
}

class ExecuteQueryArgs: Decodable {
    let language: String?
    let queryStr: String
    var parameters: [String: JSONValue]? = nil
}

class CreateFtsIndexArgs: Decodable {
    let collection: String
    let indexName: String
    let field: String
}

class StartReplicationArgs: Decodable {
    let url: String
    let collection: String
    var direction: String? = "both"
    var username: String? = nil
    var password: String? = nil
    var sessionId: String? = nil
    var cookieName: String? = nil
}

class SaveBlobArgs: Decodable {
    let dataB64: String
    var contentType: String? = "application/octet-stream"
}

class GetBlobArgs: Decodable {
    let digest: String
}

// ---------------------------------------------------------------------------

class CblitePlugin: Plugin {

    private var database: Database?
    private var replicator: Replicator?
    private var collectionListenerTokens: [ListenerToken] = []
    private var replListenerToken: ListenerToken? = nil

    // ── helpers ──────────────────────────────────────────────────────────────

    /// "scope.coll" or bare "coll" (=> _default scope), creating if needed.
    private func resolveCollection(_ db: Database, _ spec: String) throws -> Collection {
        let parts = spec.split(separator: ".", maxSplits: 1).map(String.init)
        let (scope, coll) = parts.count == 2 ? (parts[0], parts[1]) : ("_default", parts[0])
        return try db.createCollection(name: coll, scope: scope)
    }

    private func jsonStringToDict(_ s: String) throws -> [String: Any] {
        guard let data = s.data(using: .utf8),
              let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            throw NSError(domain: "cblite", code: 1,
                          userInfo: [NSLocalizedDescriptionKey: "invalid JSON from CBL"])
        }
        return obj
    }

    private func closeCurrentDatabase() throws {
        for t in collectionListenerTokens { t.remove() }
        collectionListenerTokens.removeAll()
        replListenerToken?.remove()
        replListenerToken = nil
        replicator?.stop()
        replicator = nil
        try database?.close()
        database = nil
    }

    /// One-shot config.json import, mirroring the Kotlin plugin: reads
    /// Documents/config.json, saves it to _default.config/app-config, then
    /// deletes the file. On the Simulator the Documents directory is a plain
    /// host directory, so this doubles as a config-injection mechanism.
    private func importConfigIfPresent(_ db: Database) {
        do {
            guard let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first
            else { return }
            let file = docs.appendingPathComponent("config.json")
            guard FileManager.default.fileExists(atPath: file.path) else { return }
            let json = try String(contentsOf: file, encoding: .utf8)
            let coll = try db.createCollection(name: "config", scope: "_default")
            let doc = MutableDocument(id: "app-config")
            try doc.setJSON(json)
            try coll.save(document: doc)
            try FileManager.default.removeItem(at: file)
            Logger.info("cblite: imported config.json into _default.config/app-config")
        } catch {
            Logger.error("cblite: importConfigIfPresent failed: \(error)")
        }
    }

    // ── openDatabase ─────────────────────────────────────────────────────────

    @objc public func openDatabase(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(OpenDatabaseArgs.self)
        do {
            try closeCurrentDatabase()

            var config = DatabaseConfiguration()
            config.directory = args.path
            let db = try Database(name: args.name, config: config)
            database = db

            for spec in args.collections ?? [] {
                let collection = try resolveCollection(db, spec)
                let token = collection.addChangeListener(listener: { [weak self] (change: CollectionChange) in
                    let payload = JSTypes.coerceDictionaryToJSObject(
                        ["docIds": change.documentIDs]) ?? [:]
                    DispatchQueue.main.async {
                        self?.trigger("cblite://collection-changed", data: payload)
                    }
                })
                collectionListenerTokens.append(token)
            }

            importConfigIfPresent(db)
            invoke.resolve()
        } catch {
            invoke.reject("\(error.localizedDescription)")
        }
    }

    // ── closeDatabase ────────────────────────────────────────────────────────

    @objc public func closeDatabase(_ invoke: Invoke) throws {
        do {
            try closeCurrentDatabase()
            invoke.resolve()
        } catch {
            invoke.reject("\(error.localizedDescription)")
        }
    }

    // ── getDocument ──────────────────────────────────────────────────────────

    @objc public func getDocument(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(DocArgs.self)
        guard let db = database else { return invoke.reject("Database not open") }
        do {
            let coll = try resolveCollection(db, args.collection)
            guard let doc = try coll.document(id: args.docId) else {
                invoke.resolve([:] as JSObject)
                return
            }
            let dict = try jsonStringToDict(doc.toJSON())
            invoke.resolve(JSTypes.coerceDictionaryToJSObject(dict) ?? [:])
        } catch {
            invoke.reject("\(error.localizedDescription)")
        }
    }

    // ── saveDocument ─────────────────────────────────────────────────────────

    @objc public func saveDocument(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(SaveDocumentArgs.self)
        guard let db = database else { return invoke.reject("Database not open") }
        do {
            let coll = try resolveCollection(db, args.collection)

            // `_deleted` / `__deleted` are soft-delete sentinels — purge instead.
            if case .object(let o) = args.body {
                let deleted = { (k: String) -> Bool in
                    if case .bool(true)? = o[k] { return true }
                    return false
                }
                if deleted("_deleted") || deleted("__deleted") {
                    try? coll.purge(id: args.docId)
                    invoke.resolve()
                    return
                }
            }

            let data = try JSONSerialization.data(withJSONObject: args.body.anyValue)
            let json = String(data: data, encoding: .utf8) ?? "{}"
            let doc = MutableDocument(id: args.docId)
            try doc.setJSON(json)
            try coll.save(document: doc)
            invoke.resolve()
        } catch {
            invoke.reject("\(error.localizedDescription)")
        }
    }

    // ── executeQuery ─────────────────────────────────────────────────────────

    @objc public func executeQuery(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(ExecuteQueryArgs.self)
        guard let db = database else { return invoke.reject("Database not open") }
        let language = args.language ?? "N1QL"
        guard language.uppercased() == "N1QL" else {
            return invoke.reject("Only N1QL query language is supported on iOS")
        }
        do {
            let query = try db.createQuery(args.queryStr)

            if let params = args.parameters {
                let p = Parameters()
                for (k, v) in params {
                    switch v {
                    case .string(let s): _ = p.setString(s, forName: k)
                    case .bool(let b): _ = p.setBoolean(b, forName: k)
                    case .number(let n):
                        if n.truncatingRemainder(dividingBy: 1) == 0 {
                            _ = p.setInt64(Int64(n), forName: k)
                        } else {
                            _ = p.setDouble(n, forName: k)
                        }
                    case .null: _ = p.setValue(nil, forName: k)
                    default: _ = p.setString("\(v.anyValue)", forName: k)
                    }
                }
                query.parameters = p
            }

            var rows: [[String: Any]] = []
            for result in try query.execute() {
                rows.append(try jsonStringToDict(result.toJSON()))
            }
            invoke.resolve(JSTypes.coerceDictionaryToJSObject(["rows": rows]) ?? [:])
        } catch {
            invoke.reject("\(error.localizedDescription)")
        }
    }

    // ── createFtsIndex ───────────────────────────────────────────────────────

    @objc public func createFtsIndex(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(CreateFtsIndexArgs.self)
        guard let db = database else { return invoke.reject("Database not open") }
        do {
            let coll = try resolveCollection(db, args.collection)
            let config = FullTextIndexConfiguration([args.field])
            try coll.createIndex(withName: args.indexName, config: config)
            invoke.resolve()
        } catch {
            invoke.reject("\(error.localizedDescription)")
        }
    }

    // ── startReplication ─────────────────────────────────────────────────────

    @objc public func startReplication(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(StartReplicationArgs.self)
        guard let db = database else { return invoke.reject("Database not open") }
        guard let url = URL(string: args.url) else { return invoke.reject("invalid url: \(args.url)") }
        do {
            DispatchQueue.main.async { [weak self] in
                self?.trigger("cblite://replication-status", data: ["activity": "Connecting"])
            }

            replListenerToken?.remove()
            replListenerToken = nil
            replicator?.stop()
            replicator = nil

            let coll = try resolveCollection(db, args.collection)
            let collConfig = CollectionConfiguration(collection: coll)
            var config = ReplicatorConfiguration(collections: [collConfig], target: URLEndpoint(url: url))
            switch args.direction ?? "both" {
            case "push": config.replicatorType = .push
            case "pull": config.replicatorType = .pull
            default: config.replicatorType = .pushAndPull
            }
            if let sessionId = args.sessionId, !sessionId.isEmpty {
                config.authenticator = SessionAuthenticator(
                    sessionID: sessionId, cookieName: args.cookieName ?? "SyncGatewaySession")
            } else if let u = args.username, let pw = args.password, !u.isEmpty, !pw.isEmpty {
                config.authenticator = BasicAuthenticator(username: u, password: pw)
            }
            config.continuous = true
            config.heartbeat = 30
            // Allow pulling documents created by other devices.
            config.enableAutoPurge = false

            let repl = Replicator(config: config)
            replicator = repl

            replListenerToken = repl.addChangeListener { [weak self] (change: ReplicatorChange) in
                let label: String
                switch change.status.activity {
                case .stopped: label = "Stopped"
                case .offline: label = "Offline"
                case .connecting: label = "Connecting"
                case .idle: label = "Idle"
                case .busy: label = "Busy"
                @unknown default: label = "Unknown"
                }
                var payload: JSObject = ["activity": label]
                if let err = change.status.error {
                    payload["error"] = "\(err.localizedDescription)"
                }
                DispatchQueue.main.async {
                    self?.trigger("cblite://replication-status", data: payload)
                }
            }

            repl.start()
            invoke.resolve()
        } catch {
            invoke.reject("\(error.localizedDescription)")
        }
    }

    // ── stopReplication ──────────────────────────────────────────────────────

    @objc public func stopReplication(_ invoke: Invoke) throws {
        replListenerToken?.remove()
        replListenerToken = nil
        replicator?.stop()
        replicator = nil
        invoke.resolve()
    }

    // ── saveBlob ─────────────────────────────────────────────────────────────

    @objc public func saveBlob(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(SaveBlobArgs.self)
        guard let db = database else { return invoke.reject("Database not open") }
        guard let data = Data(base64Encoded: args.dataB64) else {
            return invoke.reject("invalid base64 data")
        }
        do {
            let blob = Blob(contentType: args.contentType ?? "application/octet-stream", data: data)
            try db.saveBlob(blob: blob)
            guard let digest = blob.digest else {
                return invoke.reject("Blob has no digest after save")
            }
            invoke.resolve(["value": digest] as JSObject)
        } catch {
            invoke.reject("\(error.localizedDescription)")
        }
    }

    // ── getBlobData ──────────────────────────────────────────────────────────

    @objc public func getBlobData(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(GetBlobArgs.self)
        guard let db = database else { return invoke.reject("Database not open") }
        do {
            let props: [String: Any] = ["@type": "blob", "digest": args.digest]
            guard let blob = try db.getBlob(properties: props), let content = blob.content else {
                return invoke.reject("Blob not found: \(args.digest)")
            }
            invoke.resolve(["value": content.base64EncodedString()] as JSObject)
        } catch {
            invoke.reject("\(error.localizedDescription)")
        }
    }

    // ── predictive models (desktop-only feature) ─────────────────────────────

    @objc public func registerPredictiveModel(_ invoke: Invoke) throws {
        invoke.reject("Predictive models are not supported on iOS")
    }

    @objc public func unregisterPredictiveModel(_ invoke: Invoke) throws {
        invoke.reject("Predictive models are not supported on iOS")
    }
}

@_cdecl("init_plugin_cblite")
func initPluginCblite() -> Plugin {
    return CblitePlugin()
}

File: android/src/main/kotlin/com/plugin/cblite/CblitePlugin.kt
package com.plugin.cblite

import android.app.Activity
import android.util.Base64
import app.tauri.annotation.Command
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSArray
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import com.couchbase.lite.BasicAuthenticator
import com.couchbase.lite.Blob
import com.couchbase.lite.CouchbaseLite
import com.couchbase.lite.Collection as CblCollection
import com.couchbase.lite.Database
import com.couchbase.lite.DatabaseConfiguration
import com.couchbase.lite.ListenerToken
import com.couchbase.lite.MutableDocument
import com.couchbase.lite.Parameters
import com.couchbase.lite.Replicator
import com.couchbase.lite.ReplicatorActivityLevel
import com.couchbase.lite.ReplicatorConfiguration
import com.couchbase.lite.ReplicatorType
import com.couchbase.lite.CollectionConfiguration
import android.util.Log
import com.couchbase.lite.FullTextIndexConfiguration
import com.couchbase.lite.SessionAuthenticator
import com.couchbase.lite.URLEndpoint
import java.net.URI

@TauriPlugin
class CblitePlugin(private val activity: Activity) : Plugin(activity) {

    init {
        CouchbaseLite.init(activity)
    }

    private var database: Database? = null
    @Volatile private var replicator: Replicator? = null
    private val collectionListenerTokens: MutableList<ListenerToken> = mutableListOf()
    private var replListenerToken: ListenerToken? = null

    // ── open_database ─────────────────────────────────────────────────────────

    @Command
    fun openDatabase(invoke: Invoke) {
        val args = invoke.getArgs()
        val path = args.optString("path").takeIf { it.isNotEmpty() }
            ?: return invoke.reject("path is required", null as JSObject?)
        val name = args.optString("name").takeIf { it.isNotEmpty() }
            ?: return invoke.reject("name is required", null as JSObject?)

        try {
            closeCurrentDatabase()

            val config = DatabaseConfiguration().apply { directory = path }
            val db = Database(name, config)
            database = db

            // For each requested collection: ensure it exists and register a change listener.
            val collectionsArg = args.getJSONArray("collections")
            val collectionNames: List<String> = if (collectionsArg != null) {
                (0 until collectionsArg.length()).mapNotNull { collectionsArg.optString(it) }
            } else {
                emptyList()
            }

            for (collSpec in collectionNames) {
                val dotIdx = collSpec.indexOf('.')
                val (scope, coll) = if (dotIdx >= 0) {
                    collSpec.substring(0, dotIdx) to collSpec.substring(dotIdx + 1)
                } else {
                    "_default" to collSpec
                }
                val collection = db.createCollection(coll, scope)
                val token = collection.addChangeListener { change ->
                    val arr = JSArray()
                    change.documentIDs.forEach { arr.put(it) }
                    val obj = JSObject()
                    obj.put("docIds", arr)
                    activity.runOnUiThread { trigger("cblite://collection-changed", obj) }
                }
                collectionListenerTokens.add(token)
            }

            // Auto-import config.json from external storage (one-shot, deleted after import).
            // This runs here because Kotlin has reliable access to external storage while
            // Rust std::fs cannot access /sdcard paths on Android.
            importConfigIfPresent(db)

            invoke.resolve()
        } catch (e: Throwable) {
            invoke.reject(e.message ?: e.toString(), null as JSObject?)
        }
    }

    /** Reads config.json from external files dir, saves it to CBL, then deletes the file. */
    private fun importConfigIfPresent(db: Database) {
        try {
            val extDir = activity.getExternalFilesDir(null) ?: return
            val file = java.io.File(extDir, "config.json")
            if (!file.exists()) return

            val json = file.readText()
            // Parse and save to _default.config/app-config
            val parsed = org.json.JSONObject(json)
            val coll = db.createCollection("config", "_default")
            val doc = MutableDocument("app-config")
            doc.setJSON(parsed.toString())
            coll.save(doc)
            file.delete()
        } catch (e: Exception) {
            android.util.Log.e("CblitePlugin", "importConfigIfPresent error: ${e.message}", e)
        }
    }

    // ── create_fts_index ──────────────────────────────────────────────────────

    /** Creates (or idempotently ensures) an FTS index on a collection field. */
    @Command
    fun createFtsIndex(invoke: Invoke) {
        val args = invoke.getArgs()
        val collectionSpec = args.optString("collection").takeIf { it.isNotEmpty() }
            ?: return invoke.reject("collection is required", null as JSObject?)
        val indexName = args.optString("indexName").takeIf { it.isNotEmpty() }
            ?: return invoke.reject("indexName is required", null as JSObject?)
        val fieldArg = args.optString("field").takeIf { it.isNotEmpty() }
            ?: return invoke.reject("field is required", null as JSObject?)
        val fields = fieldArg.trim().split("\\s+".toRegex()).filter { it.isNotBlank() }
        try {
            val db = database ?: return invoke.reject("Database not open", null as JSObject?)
            val dotIdx = collectionSpec.indexOf('.')
            val (scope, coll) = if (dotIdx >= 0) {
                collectionSpec.substring(0, dotIdx) to collectionSpec.substring(dotIdx + 1)
            } else {
                "_default" to collectionSpec
            }
            val collection = db.createCollection(coll, scope)
            android.util.Log.d("CblitePlugin", "createFtsIndex: $indexName on $scope.$coll fields=$fields")
            collection.createIndex(indexName, FullTextIndexConfiguration(*fields.toTypedArray()))
            val after: Set<String> = collection.indexes
            android.util.Log.d("CblitePlugin", "createFtsIndex done, indexes now: $after")
            invoke.resolve()
        } catch (e: Throwable) {
            android.util.Log.e("CblitePlugin", "createFtsIndex FAILED: ${e.message}", e)
            invoke.reject(e.message ?: e.toString(), null as JSObject?)
        }
    }

    /** Returns the names of all indexes on a collection (used to verify FTS creation). */
    @Command
    fun listIndexes(invoke: Invoke) {
        val args = invoke.getArgs()
        val collectionSpec = args.optString("collection").takeIf { it.isNotEmpty() }
            ?: return invoke.reject("collection is required", null as JSObject?)
        try {
            val db = database ?: return invoke.reject("Database not open", null as JSObject?)
            val dotIdx = collectionSpec.indexOf('.')
            val (scope, coll) = if (dotIdx >= 0) {
                collectionSpec.substring(0, dotIdx) to collectionSpec.substring(dotIdx + 1)
            } else {
                "_default" to collectionSpec
            }
            val collection = db.createCollection(coll, scope)
            val indexNames: Set<String> = collection.indexes
            android.util.Log.d("CblitePlugin", "listIndexes $scope.$coll: $indexNames")
            val arr = JSArray()
            for (n in indexNames) { arr.put(n) }
            val result = JSObject()
            result.put("indexes", arr)
            invoke.resolve(result)
        } catch (e: Throwable) {
            invoke.reject(e.message ?: e.toString(), null as JSObject?)
        }
    }

    // ── write_export_file ─────────────────────────────────────────────────────

    /** Writes a text file to the app's external files dir (ADB-accessible). */
    @Command
    fun writeExportFile(invoke: Invoke) {
        val args = invoke.getArgs()
        val filename = args.optString("filename").takeIf { it.isNotEmpty() }
            ?: return invoke.reject("filename is required", null as JSObject?)
        val data = args.optString("data").takeIf { it.isNotEmpty() }
            ?: return invoke.reject("data is required", null as JSObject?)
        try {
            val extDir = activity?.getExternalFilesDir(null)
                ?: return invoke.reject("External storage unavailable", null as JSObject?)
            val file = java.io.File(extDir.absolutePath, filename)
            file.parentFile?.mkdirs()
            file.writeText(data)
            val result = JSObject()
            result.put("value", file.absolutePath)
            invoke.resolve(result)
        } catch (e: Throwable) {
            invoke.reject(e.message ?: e.toString(), null as JSObject?)
        }
    }

    // ── close_database ────────────────────────────────────────────────────────

    @Command
    fun closeDatabase(invoke: Invoke) {
        try {
            closeCurrentDatabase()
            invoke.resolve()
        } catch (e: Throwable) {
            invoke.reject(e.message ?: e.toString(), null as JSObject?)
        }
    }

    private fun closeCurrentDatabase() {
        collectionListenerTokens.forEach { it.close() }
        collectionListenerTokens.clear()
        replListenerToken?.close()
        replListenerToken = null
        replicator?.stop()
        replicator = null
        database?.close()
        database = null
    }

    // ── get_document ──────────────────────────────────────────────────────────

    @Command
    fun getDocument(invoke: Invoke) {
        val args = invoke.getArgs()
        val collection = args.optString("collection").takeIf { it.isNotEmpty() }
            ?: return invoke.reject("collection is required", null as JSObject?)
        val docId = args.optString("docId").takeIf { it.isNotEmpty() }
            ?: return invoke.reject("docId is required", null as JSObject?)

        try {
            val db = database ?: return invoke.reject("Database not open", null as JSObject?)
            val coll = resolveCollection(db, collection)
            val doc = coll.getDocument(docId)
            if (doc == null) {
                invoke.resolve(JSObject())
            } else {
                val json = doc.toJSON() ?: "{}"
                invoke.resolve(JSObject(json))
            }
        } catch (e: Throwable) {
            invoke.reject(e.message ?: e.toString(), null as JSObject?)
        }
    }

    // ── save_document ─────────────────────────────────────────────────────────

    @Command
    fun saveDocument(invoke: Invoke) {
        val args = invoke.getArgs()
        val collection = args.optString("collection").takeIf { it.isNotEmpty() }
            ?: return invoke.reject("collection is required", null as JSObject?)
        val docId = args.optString("docId").takeIf { it.isNotEmpty() }
            ?: return invoke.reject("docId is required", null as JSObject?)
        val body = args.getJSObject("body")
            ?: return invoke.reject("body is required", null as JSObject?)

        try {
            val db = database ?: return invoke.reject("Database not open", null as JSObject?)
            val coll = resolveCollection(db, collection)
            // `{ _deleted: true }` or `{ __deleted: true }` are soft-delete sentinels — purge the document.
            // `_deleted` is a CBL reserved property and cannot be stored in a document body.
            // `__deleted` is the preferred non-reserved tombstone used by JS deleteKnowledgeChunk.
            if (body.optBoolean("_deleted", false) || body.optBoolean("__deleted", false)) {
                // Ignore NotFound — if the doc doesn't exist, the desired state is achieved.
                try { coll.purge(docId) } catch (_: Exception) {}
                invoke.resolve()
                return
            }
            val doc = MutableDocument(docId as String)
            doc.setJSON(body.toString())
            coll.save(doc)
            invoke.resolve()
        } catch (e: Throwable) {
            invoke.reject(e.message ?: e.toString(), null as JSObject?)
        }
    }

    // ── execute_query ─────────────────────────────────────────────────────────

    @Command
    fun executeQuery(invoke: Invoke) {
        val args = invoke.getArgs()
        val language = args.optString("language").takeIf { it.isNotEmpty() } ?: "N1QL"
        val queryStr = args.optString("queryStr").takeIf { it.isNotEmpty() }
            ?: return invoke.reject("queryStr is required", null as JSObject?)
        val parameters = args.getJSObject("parameters")

        try {
            val db = database ?: return invoke.reject("Database not open", null as JSObject?)

            if (!language.equals("N1QL", ignoreCase = true)) {
                return invoke.reject("Only N1QL query language is supported on Android", null as JSObject?)
            }

            val query = db.createQuery(queryStr)

            if (parameters != null) {
                val params = Parameters()
                val keys = parameters.keys()
                while (keys.hasNext()) {
                    val key = keys.next()
                    when (val v = parameters.get(key)) {
                        is String  -> params.setString(key, v)
                        is Int     -> params.setInt(key, v)
                        is Long    -> params.setLong(key, v)
                        is Double  -> params.setDouble(key, v)
                        is Boolean -> params.setBoolean(key, v)
                        else       -> params.setString(key, v?.toString() ?: "")
                    }
                }
                query.parameters = params
            }

            val resultSet = query.execute()
            val rows = JSArray()
            for (result in resultSet) {
                val json = result.toJSON() ?: "{}"
                rows.put(JSObject(json))
            }

            val response = JSObject()
            response.put("rows", rows)
            invoke.resolve(response)
        } catch (e: Throwable) {
            invoke.reject(e.message ?: e.toString(), null as JSObject?)
        }
    }

    // ── start_replication ─────────────────────────────────────────────────────

    @Command
    fun startReplication(invoke: Invoke) {
        val args = invoke.getArgs()
        val url        = args.optString("url").takeIf { it.isNotEmpty() }
            ?: return invoke.reject("url is required", null as JSObject?)
        val collection = args.optString("collection").takeIf { it.isNotEmpty() }
            ?: return invoke.reject("collection is required", null as JSObject?)
        val direction  = args.optString("direction").takeIf { it.isNotEmpty() } ?: "both"
        val username   = args.optString("username").takeIf { it.isNotEmpty() }
        val password   = args.optString("password").takeIf { it.isNotEmpty() }
        val sessionId  = args.optString("sessionId").takeIf { it.isNotEmpty() }
        val cookieName = args.optString("cookieName").takeIf { it.isNotEmpty() }

        try {
            val db = database ?: return invoke.reject("Database not open", null as JSObject?)

            // Initial connecting status event
            val obj = JSObject()
            obj.put("activity", "Connecting")
            activity.runOnUiThread { trigger("cblite://replication-status", obj) }

            replListenerToken?.close()
            replListenerToken = null
            replicator?.stop()
            replicator = null

            val endpoint = URLEndpoint(URI(url))
            // CBL 4.0: CollectionConfiguration takes the Collection; constructor takes list + endpoint.
            // Each collection gets its own CollectionConfiguration — do NOT share instances.
            // channels = null (default) means "all channels the user has access to" per SG.
            val coll = resolveCollection(db, collection)
            val config = ReplicatorConfiguration(listOf(CollectionConfiguration(coll)), endpoint).apply {
                setType(when (direction) {
                    "push" -> ReplicatorType.PUSH
                    "pull" -> ReplicatorType.PULL
                    else   -> ReplicatorType.PUSH_AND_PULL
                })
                setAuthenticator(when {
                    !sessionId.isNullOrEmpty() ->
                        SessionAuthenticator(sessionId, cookieName ?: "SyncGatewaySession")
                    !username.isNullOrEmpty() && !password.isNullOrEmpty() ->
                        BasicAuthenticator(username, password.toCharArray())
                    else -> null
                })
                setContinuous(true)
                // Enable heartbeat to keep connection alive
                heartbeat = 30
                // ✅ CRITICAL: Disable auto purge to allow pulling documents created by OTHER DEVICES
                setAutoPurgeEnabled(false)
            }

            // Assign FIRST to prevent GC collection while starting - critical on Android!
            replicator = Replicator(config)

            replListenerToken = replicator!!.addChangeListener { change ->
                val label = when (change.status.activityLevel) {
                    ReplicatorActivityLevel.STOPPED     -> "Stopped"
                    ReplicatorActivityLevel.OFFLINE     -> "Offline"
                    ReplicatorActivityLevel.CONNECTING  -> "Connecting"
                    ReplicatorActivityLevel.IDLE        -> "Idle"
                    ReplicatorActivityLevel.BUSY        -> "Busy"
                    else                                -> "Unknown"
                }
                val obj = JSObject()
                obj.put("activity", label)
                val err = change.status.error
                if (err != null) obj.put("error", err.message ?: err.toString())
                activity.runOnUiThread { trigger("cblite://replication-status", obj) }
            }
            
            replicator!!.start()

            invoke.resolve()
        } catch (e: Throwable) {
            invoke.reject(e.message ?: e.toString(), null as JSObject?)
        }
    }

    // ── stop_replication ──────────────────────────────────────────────────────

    @Command
    fun stopReplication(invoke: Invoke) {
        replListenerToken?.close()
        replListenerToken = null
        replicator?.stop()
        replicator = null
        invoke.resolve()
    }

    // ── save_blob ─────────────────────────────────────────────────────────────

    @Command
    fun saveBlob(invoke: Invoke) {
        val args = invoke.getArgs()
        val dataB64     = args.optString("dataB64").takeIf { it.isNotEmpty() }
            ?: return invoke.reject("dataB64 is required", null as JSObject?)
        val contentType = args.optString("contentType").takeIf { it.isNotEmpty() }
            ?: "application/octet-stream"

        try {
            val db = database ?: return invoke.reject("Database not open", null as JSObject?)
            val data = Base64.decode(dataB64 as String, Base64.NO_WRAP)
            val blob = Blob(contentType as String, data as ByteArray)
            db.saveBlob(blob)
            val digest = blob.digest() ?: return invoke.reject("Blob has no digest after save", null as JSObject?)

            val result = JSObject()
            result.put("value", digest as String)
            invoke.resolve(result)
        } catch (e: Throwable) {
            invoke.reject(e.message ?: e.toString(), null as JSObject?)
        }
    }

    // ── get_blob_data ─────────────────────────────────────────────────────────

    @Command
    fun getBlobData(invoke: Invoke) {
        val args = invoke.getArgs()
        val digest = args.optString("digest").takeIf { it.isNotEmpty() }
            ?: return invoke.reject("digest is required", null as JSObject?)

        try {
            val db = database ?: return invoke.reject("Database not open", null as JSObject?)
            val props = mapOf("@type" to "blob", "digest" to digest)
            val blob = db.getBlob(props)
                ?: return invoke.reject("Blob not found: $digest", null as JSObject?)
            val content = blob.content
                ?: return invoke.reject("Blob has no content: $digest", null as JSObject?)
            val b64 = Base64.encodeToString(content, Base64.NO_WRAP)

            val result = JSObject()
            result.put("value", b64 as String)
            invoke.resolve(result)
        } catch (e: Throwable) {
            invoke.reject(e.message ?: e.toString(), null as JSObject?)
        }
    }

    // ── register_predictive_model ─────────────────────────────────────────────

    @Command
    fun registerPredictiveModel(invoke: Invoke) {
        invoke.reject("Predictive models are not supported on Android", null as JSObject?)
    }

    @Command
    fun unregisterPredictiveModel(invoke: Invoke) {
        invoke.reject("Predictive models are not supported on Android", null as JSObject?)
    }

    // ── helpers ───────────────────────────────────────────────────────────────

    private fun resolveCollection(db: Database, name: String): CblCollection {
        val dotIdx = name.indexOf('.')
        val (scope, coll) = if (dotIdx >= 0) {
            name.substring(0, dotIdx) to name.substring(dotIdx + 1)
        } else {
            "_default" to name
        }
        return db.createCollection(coll, scope)
    }
}
